File: frontend/src/api/mikrotik.ts
import { apiClient } from "@/api/client";
import type { Paginated } from "@/types";

// MikroTik RouterOS 整合。路徑帶 /api/v1 前綴（baseURL 為 /）。

export interface MikroTikRouter {
  id: string;
  name: string;
  api_url: string;
  api_username: string;
  enabled: boolean;
  verify_tls: boolean;
  sync_interval_seconds: number;
  sync_dhcp: boolean;
  sync_dhcp_ranges: boolean;
  sync_firewall: boolean;
  sync_nat: boolean;
  sync_address_lists: boolean;
  sync_vpn: boolean;
  /** 預設關：全表 ARP 在大型路由器上可能是上萬列 */
  sync_arp: boolean;
  /** CPU 超過這個百分比就停掉本輪剩下的區段 */
  cpu_load_limit: number;
  section_delay_ms: number;
  max_response_mb: number;
  description: string | null;
  scope_subnet_ids: string[] | null;
  routeros_version: string | null;
  board_name: string | null;
  last_sync_at: string | null;
  last_error: string | null;
  /** 逐區段耗時／CPU；`stopped` 代表本輪因為 CPU 超標提早停止 */
  last_cost: Record<string, unknown> | null;
}

export interface MikroTikWrite {
  name: string;
  api_url: string;
  api_username: string;
  api_password?: string;
  enabled?: boolean;
  verify_tls?: boolean;
  sync_interval_seconds?: number;
  sync_dhcp?: boolean;
  sync_dhcp_ranges?: boolean;
  sync_firewall?: boolean;
  sync_nat?: boolean;
  sync_address_lists?: boolean;
  sync_vpn?: boolean;
  sync_arp?: boolean;
  cpu_load_limit?: number;
  section_delay_ms?: number;
  max_response_mb?: number;
  description?: string;
  scope_subnet_ids?: string[];
}

/**
 * 連線診斷。比其他整合多回 `rows` 與 `seconds` —— 客戶的 MikroTik 是主力路由器，
 * 「要不要開 ARP 這一段」要由看得到數字的人決定。
 */
export interface MikroTikDiagnosis {
  api_url: string;
  version: string | null;
  board_name: string | null;
  identity: string | null;
  cpu_load: number | null;
  cpu_load_after?: number | null;
  free_memory: number | null;
  total_memory: number | null;
  uptime: string | null;
  ok_count: number;
  checks: {
    endpoint: string; ok: boolean; rows?: number; seconds?: number;
    absent?: boolean; error?: string;
  }[];
}

export interface MikroTikRule {
  id: string; table_name: string; chain: string | null; position: number;
  action: string | null; disabled: boolean;
  src_address: string | null; dst_address: string | null;
  protocol: string | null; src_port: string | null; dst_port: string | null;
  in_interface: string | null; out_interface: string | null;
  to_addresses: string | null; to_ports: string | null; comment: string | null;
}

export interface MikroTikAddressListEntry {
  id: string; list_name: string; address: string;
  dynamic: boolean; timeout: string | null; comment: string | null;
}

export async function listMikroTik(): Promise<Paginated<MikroTikRouter>> {
  const { data } = await apiClient.get<Paginated<MikroTikRouter>>("/api/v1/mikrotik", {
    params: { page: 1, page_size: 200 },
  });
  return data;
}

export async function createMikroTik(p: MikroTikWrite): Promise<MikroTikRouter> {
  const { data } = await apiClient.post<MikroTikRouter>("/api/v1/mikrotik", p);
  return data;
}

export async function updateMikroTik(
  id: string, p: Partial<MikroTikWrite>,
): Promise<MikroTikRouter> {
  const { data } = await apiClient.patch<MikroTikRouter>(`/api/v1/mikrotik/${id}`, p);
  return data;
}

export async function deleteMikroTik(id: string): Promise<void> {
  await apiClient.delete(`/api/v1/mikrotik/${id}`);
}

export async function testMikroTik(id: string): Promise<MikroTikDiagnosis> {
  const { data } = await apiClient.post<MikroTikDiagnosis>(`/api/v1/mikrotik/${id}/test`);
  return data;
}

export async function syncMikroTik(id: string): Promise<{ task_id: string }> {
  const { data } = await apiClient.post(`/api/v1/mikrotik/${id}/sync`);
  return data;
}

export async function listMikroTikRules(
  id: string, table?: string,
): Promise<MikroTikRule[]> {
  const { data } = await apiClient.get<{ items: MikroTikRule[] }>(
    `/api/v1/mikrotik/${id}/rules`, { params: table ? { table } : undefined });
  return data.items ?? [];
}

export async function listMikroTikAddressLists(
  id: string, listName?: string,
): Promise<MikroTikAddressListEntry[]> {
  const { data } = await apiClient.get<{ items: MikroTikAddressListEntry[] }>(
    `/api/v1/mikrotik/${id}/address-lists`,
    { params: listName ? { list_name: listName } : undefined });
  return data.items ?? [];
}

File: backend/app/schemas/mikrotik.py
"""MikroTik RouterOS 整合 schemas。

比其他整合多了一組「安全參數」（cpu_load_limit / section_delay_ms / max_response_mb）——
客戶的 MikroTik 是主力路由器，這些不是效能微調而是設定層級的保護，因此可讀可改。
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Annotated, Any

from pydantic import Field, HttpUrl

from app.schemas.base import StrictModel


class MikroTikBase(StrictModel):
    name: Annotated[str, Field(min_length=1, max_length=128)]
    api_url: HttpUrl
    api_username: Annotated[str, Field(min_length=1, max_length=128)]
    enabled: bool = True
    verify_tls: bool = True
    # 主力路由器 → 預設 900 秒（其他整合是 300）
    sync_interval_seconds: Annotated[int, Field(ge=60, le=86400)] = 900

    sync_dhcp: bool = True
    sync_dhcp_ranges: bool = True
    sync_firewall: bool = True
    sync_nat: bool = True
    sync_address_lists: bool = True
    sync_vpn: bool = True
    #: ⚠️ 預設關：全表 ARP 在大型路由器上可能是上萬列。
    #: 先用「測試連線」看它回報幾列、花幾秒，再自己決定要不要開。
    sync_arp: bool = False

    cpu_load_limit: Annotated[int, Field(ge=0, le=100)] = 70
    section_delay_ms: Annotated[int, Field(ge=0, le=10000)] = 300
    max_response_mb: Annotated[int, Field(ge=1, le=256)] = 8

    description: Annotated[str | None, Field(max_length=2048)] = None
    scope_subnet_ids: list[uuid.UUID] | None = None


class MikroTikCreate(MikroTikBase):
    api_password: Annotated[str, Field(min_length=1, max_length=512)]


class MikroTikUpdate(StrictModel):
    name: Annotated[str | None, Field(min_length=1, max_length=128)] = None
    api_url: HttpUrl | None = None
    api_username: Annotated[str | None, Field(min_length=1, max_length=128)] = None
    api_password: Annotated[str | None, Field(min_length=1, max_length=512)] = None
    enabled: bool | None = None
    verify_tls: bool | None = None
    sync_interval_seconds: Annotated[int | None, Field(ge=60, le=86400)] = None
    sync_dhcp: bool | None = None
    sync_dhcp_ranges: bool | None = None
    sync_firewall: bool | None = None
    sync_nat: bool | None = None
    sync_address_lists: bool | None = None
    sync_vpn: bool | None = None
    sync_arp: bool | None = None
    cpu_load_limit: Annotated[int | None, Field(ge=0, le=100)] = None
    section_delay_ms: Annotated[int | None, Field(ge=0, le=10000)] = None
    max_response_mb: Annotated[int | None, Field(ge=1, le=256)] = None
    description: Annotated[str | None, Field(max_length=2048)] = None
    scope_subnet_ids: list[uuid.UUID] | None = None


class MikroTikRead(StrictModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    api_url: str
    api_username: str
    enabled: bool
    verify_tls: bool
    sync_interval_seconds: int
    sync_dhcp: bool
    sync_dhcp_ranges: bool
    sync_firewall: bool
    sync_nat: bool
    sync_address_lists: bool
    sync_vpn: bool
    sync_arp: bool
    cpu_load_limit: int
    section_delay_ms: int
    max_response_mb: int
    description: str | None = None
    scope_subnet_ids: list[uuid.UUID] | None = None
    routeros_version: str | None = None
    board_name: str | None = None
    last_sync_at: datetime | None = None
    last_error: str | None = None
    #: 逐區段的耗時／CPU（含「因為 CPU 超標而提早停止」的原因）
    last_cost: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime


class MikroTikRuleRead(StrictModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    router_id: uuid.UUID
    table_name: str
    chain: str | None = None
    position: int
    action: str | None = None
    disabled: bool
    src_address: str | None = None
    dst_address: str | None = None
    protocol: str | None = None
    src_port: str | None = None
    dst_port: str | None = None
    in_interface: str | None = None
    out_interface: str | None = None
    to_addresses: str | None = None
    to_ports: str | None = None
    comment: str | None = None
    synced_at: datetime


class MikroTikAddressListRead(StrictModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    router_id: uuid.UUID
    list_name: str
    address: str
    dynamic: bool
    timeout: str | None = None
    comment: str | None = None
    synced_at: datetime
